File: csharp/HearthstoneReplays/Parser/Handlers/EntityChosenHandler.cs
using System;
using System.Collections.Generic;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using HearthstoneReplays.Parser.ReplayData.Meta;

namespace HearthstoneReplays.Parser.Handlers
{
	public class EntityChosenHandler
	{
		public static void Handle(string timestamp, string data, ParserState state)
		{
			data = data.Trim();
			var match = Regexes.EntitiesChosenRegex.Match(data);
			if(match.Success)
			{
				/*NOTE: in 10357, "Player" is bugged, it's treating a player ID
				as an entity ID, resulting in "Player=GameEntity"
				For our own sanity we keep the old playerID logic from the
				previous builds, we'll change to "player" when it's fixed.*/
				var rawEntity = match.Groups[1].Value;
				var rawPlayer = match.Groups[2].Value;
				var count = int.Parse(match.Groups[3].Value);
				var entity = Helper.ParseEntity(rawEntity, state);
				var player = Helper.ParseEntity(rawPlayer, state);
				var cEntities = new ChosenEntities {Entity = entity, PlayerId = player, Count = count, Choices = new List<Choice>(), TimeStamp = timestamp};
				state.CurrentGame.Data.Add(cEntities);
				state.CurrentChosenEntites = cEntities;
				return;
			}
			match = Regexes.EntitiesChosenEntitiesRegex.Match(data);
			if(match.Success)
			{
				var index = int.Parse(match.Groups[1].Value);
				var rawEntity = match.Groups[2].Value;
				var entity = Helper.ParseEntity(rawEntity, state);
				var choice = new Choice {Entity = entity, Index = index};
				state.CurrentChosenEntites.Choices.Add(choice);
				return;
			}
			Console.WriteLine("Warning: Unhandled chosen entities: " + data);
		}
	}
}

File: java/src/main/java/info/hearthsim/hsreplay/parser/handlers/EntityChosenHandler.java
package info.hearthsim.hsreplay.parser.handlers;

import info.hearthsim.hsreplay.parser.Helper;
import info.hearthsim.hsreplay.parser.ParserState;
import info.hearthsim.hsreplay.parser.Regexes;
import info.hearthsim.hsreplay.parser.replaydata.gameactions.ChosenEntities;
import info.hearthsim.hsreplay.parser.replaydata.meta.Choice;

import java.util.ArrayList;
import java.util.regex.Matcher;

import lombok.extern.slf4j.Slf4j;

@Slf4j
public class EntityChosenHandler {

	public static void handle(String timestamp, String data, ParserState state) throws Exception {
		data = data.trim();
		Matcher match = Regexes.EntitiesChosenRegex.matcher(data);
		if (match.matches()) {
			// NOTE: in 10357, "Player" is bugged, it's treating a player ID
			// as an entity ID, resulting in "Player=GameEntity"
			// For our own sanity we keep the old playerID logic from the
			// previous builds, we'll change to "player" when it's fixed.
			String rawEntity = match.group(1);
			String rawPlayer = match.group(2);
			int count = Integer.parseInt(match.group(3));
			int entity = Helper.parseEntity(rawEntity, state);
			int player = Helper.parseEntity(rawPlayer, state);

			ChosenEntities cEntities = new ChosenEntities(timestamp, entity, player, count, new ArrayList<Choice>());
			state.getCurrentGame().getData().add(cEntities);
			state.setCurrentChosenEntities(cEntities);
			return;
		}

		match = Regexes.EntitiesChosenEntitiesRegex.matcher(data);
		if (match.matches()) {
			int index = Integer.parseInt(match.group(1));
			String rawEntity = match.group(2);
			int entity = Helper.parseEntity(rawEntity, state);
			Choice choice = new Choice(entity, index);
			state.getCurrentChosenEntities().getChoices().add(choice);
			return;
		}

		log.warn("Warning: Unhandled chosen entities: " + data);
	}

}
